### PR TITLE
Deposits information block on account overview page

### LINKED ISSUE #1410: Deposit information block for the Account Overview page
Dervied from #1339 

Implement data retrieval and display for deposit information block (details paJDYF-18j-p2 ). There may be multiple blocks, one for each settlement currency.

Feature toggle is located in `is_account_overview_page_enabled` method of `WC_Payments_Admin` class.

### PR DESCRIPTION
**Pull request by dmvrtx**
Fixes #1410 

#### Changes proposed in this Pull Request

Introduce a new block on Account Overview page: deposits information. It contains data similar to one found on the Deposits page but provides a separate block for each deposit currency.

It also contains some changes to the TypeScript tooling set-up to improve compatibility between linters provided by different tools.

Task list:

 - [x] Create a mark-up for a deposit information block
 - [x] Create a block per deposit currency once deposit data is loaded
 - [x] Extract supporting functions into a separate module in `utils`
 - [x] Add the instant deposit link
 - [x] Update/add styles to match the Figma files.
 - [x] Add more TypeScript declarations. This could easily be a separate issue. (https://github.com/Automattic/woocommerce-payments/issues/1967)

#### Testing instructions

* Checkout out server PR 837-gh-Automattic/woocommerce-payments-server if it has not already been merged.
 * Enable option _wcpay_feature_account_overview
 * Open wp-admin and go under **Payments -> Overview**
 * Ensure that there is at least one deposit information block and that number of blocks matches number of currencies set-up for deposits
 * Ensure that block design matches one in ITsL0PJ0Xz4ZzQpsWyMgiI-fi-1565%3A1798

-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

File: client/data/deposits/selectors.js
Comment (RadoslavGeorgiev) [inline client/data/deposits/selectors.js:102]: Some interesting logic here, try this snippet to see how it works:

```js
[ 'eur', 'gbp', 'bgn', 'huf', 'bam' ].sort( ( a, b ) => ( 'gbp' === b ) ? 1 : 0 )
```

Ideally the default currency should be first, and the order of the rest can remain the same.

File: includes/admin/class-wc-rest-payments-deposits-controller.php
Comment (RadoslavGeorgiev) [inline includes/admin/class-wc-rest-payments-deposits-controller.php:119]: __A note about testing:__

The best environment to test this in would be one with a WCPay account, which has multiple bank accounts with different currencies associated with it, as well as completed payouts to those accounts.

In my case I have a fresh setup with multiple bank accounts, within a fresh Stripe Connect environment, and I have to wait seven days for the payouts to happen, which I can obviously not afford. To test the UI, add this line:

```suggestion
		echo file_get_contents( WCPAY_ABSPATH . '/all-deposits-overview.json' ); exit;
		return $this->forward_request( 'get_all_deposits_overviews', [] );
```

And this file to the root of the plugin:

```json
{
    "deposit": {
        "last_paid": [
          {
            "id": "po_...",
            "date": 1619395200000,
            "type": "deposit",
            "amount": 3160,
            "status": "paid",
            "bankAccount": " ••••  (USD)",
            "currency": "eur",
            "automatic": true,
            "fee": 0,
            "fee_percentage": 0,
            "created": 1619395200
          },
          {
            "id": "po_...",
            "date": 1619395200000,
            "type": "deposit",
            "amount": 1000,
            "status": "paid",
            "bankAccount": " ••••  (EUR)",
            "currency": "gbp",
            "automatic": true,
            "fee": 0,
            "fee_percentage": 0,
            "created": 1619395200
          }
        ],
        "next_scheduled": [
            {
                "id": "wcpay_estimated_weekly_eur_1622678400",
                "date": 1622678400000,
                "type": "deposit",
                "amount": 3343,
                "status": "estimated",
                "bankAccount": null,
                "currency": "eur",
                "automatic": true,
                "fee": 0,
                "fee_percentage": 0,
                "created": 1622678400
            },
            {
                "id": "wcpay_estimated_weekly_gbp_1622678400",
                "date": 1622678400000,
                "type": "deposit",
                "amount": 1656,
                "status": "estimated",
                "bankAccount": null,
                "currency": "gbp",
                "automatic": true,
                "fee": 0,
                "fee_percentage": 0,
                "created": 1622678400
            }
        ]
    },
    "balance": {
        "pending": [
            {
                "amount": 3343,
                "currency": "eur",
                "source_types": {
                    "card": 3343
                }
            },
            {
                "amount": 1656,
                "currency": "gbp",
                "deposits_count": 2,
                "source_types": {
                    "card": 1656
                }
            }
        ],
        "available": [
            {
                "amount": 2030,
                "currency": "eur",
                "source_types": {
                    "card": 0
                }
            },
            {
                "amount": 0,
                "currency": "gbp",
                "source_types": {
                    "card": 0
                }
            }
        ]
    },
    "account": {
        "deposits_enabled": true,
        "deposits_schedule": {
            "delay_days": 7,
            "interval": "weekly",
            "weekly_anchor": "thursday"
        },
        "default_currency": "gbp"
    }
}
```

**Comment by RadoslavGeorgiev** [comment]
Note: For the last commit I needed to comment out https://github.com/Automattic/woocommerce-payments/blob/add/1410-deposits-information-on-account-overview/package.json#L64. We need to figure out how to reliably load `declarations.d.ts` because otherwise I'm getting the following errors:

```
client/components/deposits-information/index.tsx(23,42): error TS2307: Cannot find module 'data' or its corresponding type declarations.
client/components/deposits-information/index.tsx(24,22): error TS2307: Cannot find module 'components/loadable' or its corresponding type declarations.
client/components/deposits-information/index.tsx(25,31): error TS2307: Cannot find module 'components/details-link' or its corresponding type declarations.
client/components/deposits-information/index.tsx(31,8): error TS2307: Cannot find module 'deposits/overview' or its corresponding type declarations.
client/components/deposits-information/index.tsx(32,52): error TS2307: Cannot find module 'utils/currency' or its corresponding type declarations.
client/components/deposits-information/index.tsx(97,12): error TS2503: Cannot find namespace 'AccountOverview'.
client/components/deposits-information/index.tsx(98,11): error TS2503: Cannot find namespace 'AccountOverview'.
client/components/deposits-information/index.tsx(200,37): error TS2503: Cannot find namespace 'AccountOverview'.
husky - pre-commit hook exited with code 1 (error)
```

Those are not an issue during compilation 🤷

**Comment by dmvrtx** [comment]
> We need to figure out how to reliably load `declarations.d.ts`

This might to do with the fact that `lint-staged` only applies checks to staged files and if `declaration.d.ts` is not touched, it is not included. Maybe setting `incremental` to `true` in `tsconfig.json` and adding `--incremental` to the aforementioned line would do the trick, as compiler will keep a build information cache.

**Comment by RadoslavGeorgiev** [comment]
@LevinMedia could you please take a look? Except for the instant deposit button, everything should be there.

File: client/data/deposits/selectors.js
Comment (naman03malhotra) [inline client/data/deposits/selectors.js:92]: Just a thought: can we move this logic to resolver so that this computation is not done each time the selector is called, or maybe we can memorize the result of this selector. What do you think?

**Review by naman03malhotra — CHANGES_REQUESTED** [review]
The PR is looking great so far! I gave my first pass of review, I'll give it another go tomorrow. I've tested the branch as well with the fixture you provided, and it is working as expected. Left a few comments, none of which are major. Awesome work @RadoslavGeorgiev and @dmvrtx! 😃 

ps: It'd be great if we can get a second set of 👀  as well for testing and final approval. (Also in case I missed something).

**Comment by naman03malhotra** [comment]
Note: Unit tests are missing, which I think will get to review tomorrow with the Instant deposit feature.

File: client/components/deposits-information/index.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/index.tsx:97]: That's actually done here: https://github.com/Automattic/woocommerce-payments/blob/add/1410-deposits-information-on-account-overview/client/components/deposits-information/declarations.d.ts

File: client/components/deposits-information/index.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/index.tsx]: Fixed here: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/8cea60eee2b67b9f1c79e00b54a7176aaa3cbb29. It was a last-minute change and I didn't look at the console at that time. My assumption was that the attribute was simply passed to the DOM.

File: client/components/deposits-information/index.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/index.tsx:34]: Done in this commit: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/31227dcb0cc4af7a0253ee63470776c3ffa0071c

File: client/components/deposits-information/index.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/index.tsx]: Added [here](https://github.com/Automattic/woocommerce-payments/pull/1892/commits/31227dcb0cc4af7a0253ee63470776c3ffa0071c#diff-b87314bcbeb5317e5b0c8e60d1b4df677a62ffb0d910f359cc18d5523c499eb0R118), thank you for noticing!

File: client/components/deposits-information/style.scss
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/style.scss]: Fixed in https://github.com/Automattic/woocommerce-payments/pull/1892/commits/31227dcb0cc4af7a0253ee63470776c3ffa0071c. I didn't expect that the small size would perfectly match the 16px we needed :D

File: client/data/deposits/selectors.js
Comment (RadoslavGeorgiev) [reply client/data/deposits/selectors.js]: Thank you for the suggestion, but it doesn't work either. Turns out that browsers treat this differently.

It works in Chrome:

```js
[ 'eur', 'gbp', 'bgn', 'huf', 'bam' ].sort( ( a, b ) => ( 'gbp' === b ) ? 0 : -1 )
(5) ["gbp", "bam", "huf", "bgn", "eur"]
```

But not in Firefox:

```js
[ 'eur', 'gbp', 'bgn', 'huf', 'bam' ].sort( ( a, b ) => ( 'gbp' === b ) ? 0 : -1 )
Array(5) [ "eur", "gbp", "bgn", "huf", "bam" ]
```

The way I initially implemented it was the other way around.

To avoid this, I'm just filtering the array twice and concatenating the result. Not as elegant, but definitely more reliable.

File: includes/admin/class-wc-payments-admin.php
Comment (RadoslavGeorgiev) [reply includes/admin/class-wc-payments-admin.php]: Done in [this commit](https://github.com/Automattic/woocommerce-payments/pull/1892/commits/0e1f2c669ceb4205f953893272b2cc5124a9b1ca) :)

File: client/data/deposits/resolvers.js
Comment (RadoslavGeorgiev) [reply client/data/deposits/resolvers.js:80]: Short answer: We don't.

Long answer: During the build/release process, `languages/woocommerce-payments.pot` will be re-generated and uploaded along with the plugin. Translations are then done at [translate.wordpress.org](https://translate.wordpress.org/projects/wp-plugins/woocommerce-payments/stable/) and managed by WordPress itself.

Even if we are contributing translations, it is not directly done in the repo, but through that website as well.

File: webpack.config.js
Comment (RadoslavGeorgiev) [reply webpack.config.js]: See [this discussion](https://github.com/Automattic/woocommerce-payments/pull/1748#discussion_r639484200). The plan is to remove it, but not yet.

File: package.json
Comment (RadoslavGeorgiev) [reply package.json]: Based on https://github.com/TypeStrong/ts-loader/tags?after=v9.0.1, I think we can keep the caret. Version `v9.0.0` does contain a breaking change, and is a full new version, so I think we can rely on that. Do you agree?

File: client/data/deposits/selectors.js
Comment (RadoslavGeorgiev) [reply client/data/deposits/selectors.js:92]: That's good advice. We could potentially also use `useEffect` to make it a bit more React-style.

I don't think it's worth it though:
- I'd like to keep this logic in the selector, and that leaves memorization as the only choice.
- Looking at the design of the page, this selector will be called few times.
- There is not a lot of computation in the selector, and the dataset is and will be pretty small in most cases.

Do you agree?

**Comment by RadoslavGeorgiev** [comment]
@naman03malhotra I've addressed your feedback and will proceed with instant deposits and test now. In the meantime, could you please 👀 the feedback that's already in place?

File: client/components/deposits-information/test/index.js
Comment (naman03malhotra) [inline client/components/deposits-information/test/index.js:104]: Maybe we don't need a structural test for the single currency if we are testing for multiple currencies. Since the structure is repetitive. Test for multiple currency covers for both, let me know if I am missing something.

File: client/components/deposits-information/test/index.js
Comment (naman03malhotra) [inline client/components/deposits-information/test/index.js:122]: Since the structural test is already covered, maybe we can do an exact match here?

File: client/components/deposits-information/test/index.js
Comment (naman03malhotra) [inline client/components/deposits-information/test/index.js:141]: Can we just test for the instant deposit button here, instead of the whole component?

File: client/overview/index.js
Comment (naman03malhotra) [inline client/overview/index.js:16]: Nit: reorder imports

**Review by naman03malhotra — COMMENTED** [review]
Looks good so far! yet to review tests and utils migration in-depth, left a few comments which are mostly nits and test related. I'll continue with the final review tomorrow.

ps: pushed a small fix so that it unblocks David for testing.

File: client/data/deposits/selectors.js
Comment (naman03malhotra) [reply client/data/deposits/selectors.js]: I agree, make sense, though can you please add a comment on the top indicating that we can add memorization later on if required.

**Comment by LevinMedia** [comment]
@RadoslavGeorgiev Got this running this am - looking good. I haven't been able to test the instant deposits feature.

File: client/data/deposits/test/selectors.js
Comment (naman03malhotra) [inline client/data/deposits/test/selectors.js:236]: Nice!!

File: client/components/deposits-information/test/index.js
Comment (naman03malhotra) [inline client/components/deposits-information/test/index.js:94]: Maybe we don't need a structural test for the single currency if we are testing for multiple currencies. Since the structure is repetitive. Test for multiple currency covers for both, let me know if I am missing something.

File: client/deposits/utils/index.js
Comment (naman03malhotra) [inline client/deposits/utils/index.js:1]: Some tests are missing for this file as they were not present in the first place. Maybe we can include those tests in a separate PR later. I'll leave that up to you to decide.

File: client/components/deposits-information/test/index.js
Comment (naman03malhotra) [inline client/components/deposits-information/test/index.js:127]: Just a thought! can we include this for one currency in the fixture file for manual testing purposes?

**Review by naman03malhotra — CHANGES_REQUESTED** [review]
So far so good! Left small comments related to tests, none of which are major. I am yet to test it with the server PR, any instructions if any would be great.

ps: I possible we can quickly get a second set of 👀  from the team on Monday, in case I missed something.

**Comment by naman03malhotra** [comment]
Hi @LevinMedia, to enable the Instant deposits feature using fixtures, you can use the fixture file present in the test directory.

Please replace `L119` with `echo file_get_contents( WCPAY_ABSPATH . '/client/data/deposits/test/overviews.fixture.json' ); exit;`
https://github.com/Automattic/woocommerce-payments/blob/6ce5f1e2783ff5e12c5cc9d742da8be18e7b4e30/includes/admin/class-wc-rest-payments-deposits-controller.php#L118-L119

@RadoslavGeorgiev can elaborate more on how to test it with server PR.

File: client/data/deposits/selectors.js
Comment (RadoslavGeorgiev) [reply client/data/deposits/selectors.js]: Thank you!

File: client/components/deposits-information/test/index.js
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/test/index.js]: That is right, if there are multiple-currency snapshots, having the simple one is not useful. Removed in https://github.com/Automattic/woocommerce-payments/pull/1892/commits/8b483572ac099caecf6d4c201369efc03b0218bb

File: client/overview/index.js
Comment (RadoslavGeorgiev) [reply client/overview/index.js]: Done [here](https://github.com/Automattic/woocommerce-payments/pull/1892/commits/ce426cc54a41bdd915a32a2e72d89eea4cc72b97#diff-a70fc7fd837200557381243b7777d7df8134f8c183630fae8d73cf221a83f511)

File: client/components/deposits-information/test/index.js
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/test/index.js:141]: Done here: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/ae92567b54d4a5931b4f449ce6235f544afad425

I still kept the snapshot, in order to make sure that the button not only exists on the page, but is shown in the right place.

File: client/components/deposits-information/test/index.js
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/test/index.js]: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/6ac5a3d9986b600ffb8f847dd2736aead7b3a30a adds more specificity. The snapshot is gone and instead we're looking for 4 `$0.00`s and no extra information.

File: client/components/deposits-information/index.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/index.tsx]: Done in https://github.com/Automattic/woocommerce-payments/pull/1892/commits/ce426cc54a41bdd915a32a2e72d89eea4cc72b97

File: client/data/deposits/selectors.js
Comment (RadoslavGeorgiev) [reply client/data/deposits/selectors.js]: Alright. Done here: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/ce426cc54a41bdd915a32a2e72d89eea4cc72b97#diff-35978bc52ccca1b99c8528f6fe263d7c46de7754ab5a0ae54f48fcecfa0ee11fR76

**Comment by RadoslavGeorgiev** [comment]
@naman03malhotra  I addressed all but two minor comments out of your feedback (I agree, but none of them are high priority). Could you please take a (🤞) last look?

> @RadoslavGeorgiev can elaborate more on how to test it with server PR.

The server PR does not require any additional setup. Once checked out, this PR should simply work as well.

File: client/components/deposits-information/style.scss
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/style.scss]: Actually, I'm not sure why @dmvrtx picked 390px here. 480 seems like a better breakpoint for the overview page, which is why I switched to it in https://github.com/Automattic/woocommerce-payments/pull/1892/commits/cff2fe84d70a512a28e7b28db3144720bee2e37a. The possible breakpoints are declared in `client/stylesheets/abstracts/_breakpoints.scss`

File: client/deposits/utils/index.js
Comment (RadoslavGeorgiev) [reply client/deposits/utils/index.js:1]: Here they are: https://github.com/Automattic/woocommerce-payments/pull/1892/commits/fa76c4aa9c99a73a36c9e7609867b3a6217fa9e7 :)

Comment (naman03malhotra) [reply client/deposits/utils/index.js:1]: Thank you! so much for this ❤️ .

**Review by naman03malhotra — APPROVED** [review]
Gave it a final review! and we have a PR ready to 🚢 . 
🥳

File: client/components/deposits-information/loading.tsx
Comment (reykjalin) [inline client/components/deposits-information/loading.tsx:14]: I think we should avoid `any` at all costs. Can we change this to `JSX.Element`? Or a `React.Node` — I think that's the React element type, not sure though.

**Review by reykjalin — CHANGES_REQUESTED** [review]
I have some questions regarding the module declarations, and there's one `any` return type we need to fix 🙂

File: client/components/deposits-information/declarations.d.ts
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/declarations.d.ts]: I think I got confused around [here](https://github.com/Automattic/woocommerce-payments/pull/1892#issuecomment-849794172) and added them by mistake 🤔 

Anyway, your suggestion works: `i18n` in 4.1.0 includes types, and there is a separate package for `components`. Both are now in place.

File: client/components/deposits-information/loading.tsx
Comment (RadoslavGeorgiev) [reply client/components/deposits-information/loading.tsx]: Yes, `JSX.Elements` did the job, it was an oversight.

**Review by reykjalin — APPROVED** [review]
LGTM 👍 

:shipit:

File: client/overview/style.scss
Comment (dechov) [inline client/overview/style.scss:4]: Not that big a deal but FYI `<Page isNarrow>` accomplishes the same.

**Comment by LevinMedia** [comment]
@dmvrtx @RadoslavGeorgiev one small follow up issue here - Testing on develop after merge noticed if there's no pending deposits - I see a Zero under the deposit amount - We should sub that out with a text string that reads "No pending deposits" 

It also appears that the test mode notice copy has disappeared (At least while in dev mode), the only thing left is the link to the settings. 

Screen Shot 2021-05-31 at 3 16 59 PM

**Pull request merged**
